src/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -0,0 +1,194 @@
+ * nothing worked.
+ * This was "fixed" by commenting out the NVIC_EnableIRQ, stopping the interrupt
+ * from ever being enabled.
+ *
+ * For commit from before code was removed, refer to commit 81624521a8b2c4b66480193e88cf32782aaee84d.
+ */
+
+#include <EVT/io/pin.hpp>
+#include <EVT/io/platform/f4xx/ADCf4xx.hpp>
+#include <EVT/io/platform/f4xx/GPIOf4xx.hpp>
+#include <HALf4/stm32f4xx.h>
+#include <HALf4/stm32f4xx_hal_adc.h>
+
+namespace EVT::core::IO {
+
+// Init static member variables
+ADC_HandleTypeDef ADCf4xx::halADC = {0};
+Pin ADCf4xx::channels[MAX_CHANNELS];
+uint16_t ADCf4xx::buffer[MAX_CHANNELS];
+DMA_HandleTypeDef ADCf4xx::halDMA = {0};
+
+ADCf4xx::ADCf4xx(Pin pin) : ADC(pin) {
+    // Flag representing if the ADC has been configured yet
+    static bool halADCisInit = false;
+
+    // "Rank" represents the order in which the channels are added
+    // Also represents the total number of added channels
+    static uint8_t rank = 1;
+
+    // Maximum number of ADC channels have already been added
+    if (rank == MAX_CHANNELS) {
+        return;
+    }
+
+    // Initialization of the HAL ADC should only take place once since there is
+    // only one ADC device which has multiple channels supported.
+    if (halADCisInit) {
+        HAL_ADC_Stop_DMA(&halADC);
+        HAL_DMA_DeInit(&halDMA);
+    } else {
+        __HAL_RCC_DMA2_CLK_ENABLE();
+        halADCisInit = true;
+    }
+
+    addChannel(rank);
+
+    initADC(rank);
+
+    initDMA();
+
+    HAL_ADC_Start_DMA(&halADC, reinterpret_cast<uint32_t*>(&buffer[0]),
+                      rank);
+
+    rank++;
+}
+
+float ADCf4xx::read() {
+    float percentage = readPercentage();
+    return percentage * VREF_POS;
+}
+
+uint32_t ADCf4xx::readRaw() {
+    // Search through list of channels to determine which DMA buffer index to
+    // use
+    uint8_t channelNum = 0;
+    while (channels[channelNum] != pin)
+        channelNum++;
+    return buffer[channelNum];
+}
+
+float ADCf4xx::readPercentage() {
+    float raw = static_cast<float>(readRaw());
+    return static_cast<float>(raw / MAX_RAW);
+}
+
+void ADCf4xx::initADC(uint8_t num_channels) {
+    /** Configure the global features of the ADC (Clock, Resolution, Data
+     * Alignment and number of conversion)
+      */
+    halADC.Instance = ADC1;
+    halADC.Init.ClockPrescaler = ADC_CLOCK_SYNC_PCLK_DIV2;
+    halADC.Init.Resolution = ADC_RESOLUTION_12B;
+    halADC.Init.ScanConvMode = ENABLE;
+    halADC.Init.ContinuousConvMode = ENABLE;
+    halADC.Init.DiscontinuousConvMode = DISABLE;
+    halADC.Init.ExternalTrigConvEdge = ADC_EXTERNALTRIGCONVEDGE_NONE;
+    halADC.Init.ExternalTrigConv = ADC_SOFTWARE_START;
+    halADC.Init.DataAlign = ADC_DATAALIGN_RIGHT;
+    halADC.Init.NbrOfConversion = num_channels;
+    halADC.Init.DMAContinuousRequests = ENABLE;
+    halADC.Init.EOCSelection = ADC_EOC_SEQ_CONV;
+
+    __HAL_RCC_ADC1_CLK_ENABLE();
+    HAL_ADC_Init(&halADC);
+}
+
+void ADCf4xx::initDMA() {
```

Comment:
You deleted a TODO here. Do you understand what issue it was bringing up?

Reply:
I believe the todo here was find out why there was commented out code in the f3xx ADC. 
The commented code was stopping the non-DMA ADC before initializing the DMA, but that was redundant and possibly harmful as right before initDMA is called it stops DMA ADC if it was on.